qutip/tests/conftest.py

```diff
@@ -0,0 +1,197 @@
+    if verbose:
+        for i, a in enumerate(arrays):
+            if isinstance(a, np.ndarray):
+                # precision argument is only needed if the objects are ndarrays
+                r_func = functools.partial(np.core.array_repr,
+                                           precision=precision)
+            else:
+                r_func = repr
+
+            try:
+                # [diff] Remove truncation threshold from array output.
+                with np.printoptions(threshold=np.inf):
+                    r = r_func(a)
+            except Exception as exc:
+                r = '[repr failed for <{}>: {}]'.format(type(a).__name__, exc)
+            # [diff] The original truncates the output to 3 lines here.
+            msg.append(' %s: %s' % (names[i], r))
+    return '\n'.join(msg)
+
+
+# Find the private module used by numpy to store its testing utility functions
+# so that we can monkeypatch the error messages to be more verbose.  QuTiP
+# supports numpy from 1.12 upwards, so we have to search.
+_numpy_private_utils_paths = [
+    ['_private', 'utils'],    # 1.15.0 <= x
+    ['nose_tools', 'utils'],  # 1.14.0 <= x < 1.15.0
+    ['utils'],                # 1.14.0 > x
+]
+for possible_path in _numpy_private_utils_paths:
+    try:
+        module = np.testing
+        for submodule in possible_path:
+            module = getattr(module, submodule)
+        _numpy_private_utils = module
+        break
+    except NameError:
+        pass
+else:
+    # If we can't locate it for some reason, then we don't attempt to patch.
+    _numpy_private_utils = None
+
+if _numpy_private_utils is not None:
+    @pytest.fixture(autouse=True)
+    def do_not_truncate_numpy_output(monkeypatch):
+        """
+        Monkeypatch the internal numpy function used for printing arrays so
+        that we get full output that isn't cut off after three lines.
+        """
+        with monkeypatch.context() as patch:
+            patch.setattr(np.testing._private.utils, "build_err_msg",
+                          _patched_build_err_msg)
+            # Yield inside context manager just to minimize the amount of time
+            # we've monkeypatched such a core library (and a private function!)
+            yield
+
+
+@pytest.fixture
+def in_temporary_directory():
+    """
+    Creates a temporary directory for the lifetime of the fixture and changes
+    into it.  All relative paths used will be in the temporary directory, and
+    everything will automatically be cleaned up at the end of the fixture's
+    life.
+    """
+    previous_dir = os.getcwd()
+    with tempfile.TemporaryDirectory() as temporary_dir:
+        os.chdir(temporary_dir)
+        yield
+        # pytest should catch exceptions occuring in functions using the
+        # fixture, so this should always be called.  We want it here rather
+        # than outside to prevent the case of the directory failing to be
+        # removed because it is 'busy'.
+        os.chdir(previous_dir)
```

Comment:
What goes after yield will be executed even if there are Exceptions in the test functions. It is like the `finally` in Python.

Reply:
I don't think this is right - coroutines don't modify the flow of exception handling.  This is a mocked up and simplified version of `pytest` running a test:
```python
def fixture():
    """A possible fixture"""
    print("Initialising fixture.")
    yield 12
    print("Finalising fixture.")

def test(f):
    """A test case"""
    if f == 12:
        raise ValueError()

def run_tests(finalise=True):
    """MWE of how pytest runs tests."""
    fixtures = [fixture()]  # Store all fixture functions
    args = [next(f) for f in fixtures]  # Advance coroutines to get values
    successes = 0
    caught = []
    try:
        print("Running tests.")
        test(*args)
        successes += 1
    except Exception as e:  # Catch Assertion/WhateverError
        caught.append(e)
    finally:
        if finalise:
            for f in fixtures:
                # Try to finish the coroutine, i.e. control will start from the
                # `yield` statement, and this is when the finaliser is run.  If
                # it raises StopIteration that means it completed.  If it
                # doesn't, that means it must have tried to return a new value.
                try:
                    next(f)
                except StopIteration:
                    pass
                else:
                    raise Exception("fixture tried to yield more than once")
    return successes, caught
```
Now note that `run_tests(finalise=True)` will print "Finalising fixture.", but `run_tests(finalise=False)` won't.

In fact, it's not really anything to do with exceptions - it's just that control doesn't return to the coroutine until someone passes it back.  In this case, `pytest` does (by calling `next` again), so it does finalise correctly.

Reply:
Aha, this is an interesting use of `yield`. I find it quite similar to a decorator, or a context manager. You wrap the function with it every time you detect the marker.

Reply:
Yeah, it's pretty funky.  We don't yield anything here because we don't actually care about providing a value, we only care about ensuring that the setup/teardown code is executed before and after the relevant testing scopes.